### PR TITLE
PWN-7169 - swap presenter, mapping, autosize widget

### PR DESCRIPTION
**Pull request by NikolaiEmelianov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7169

## Description of Work

swap presenter, mapping, autosize widget

### CODE REVIEW

File: ui-kit/src/main/kotlin/org/p2p/uikit/components/UiKitSliderSolidButton.kt [hunk 1]
```diff
@@ -93,6 +94,11 @@ class UiKitSliderSolidButton @JvmOverloads constructor(
         textViewActionTop.text = actionText
     }
 
+    fun setActionText(actionText: TextContainer) = with(binding) {
+        actionText.applyTo(textViewAction)
```
Comment (gslevinkov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/components/UiKitSliderSolidButton.kt:98]: can we make extensions to use this method like
textViewAction.bind(actionText) 
?

```
TextView.bind(container: TextContainer) { 
    actionText.applyTo(this)
}
```
to unify working with Cell models and Containers

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidgetModel.kt [hunk 2]
```diff
@@ -43,13 +25,13 @@ sealed interface SwapWidgetModel {
     ) : SwapWidgetModel
 }
 
-private fun bigSkeleton(): SkeletonCellModel = SkeletonCellModel(
+fun bigSkeleton(): SkeletonCellModel = SkeletonCellModel(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidgetModel.kt:28]: are you sure we need  to make bigSkeleton public? it's very ambiguous

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidget.kt [hunk 3]
```diff
@@ -24,13 +30,17 @@ class SwapWidget @JvmOverloads constructor(
...
+    var onAmountChanged: (newAmount: String) -> Unit = {}
 
     init {
         minHeight = 120.toPx()
         background = shapeDrawable(shapeRounded16dp())
         backgroundTintList = backgroundTint()
         binding.textViewWidgetTitle.setTextColor(widgetTitleTint())
         initInputType = binding.editTextAmount.inputType
+        binding.editTextAmount.doAfterTextChanged { resizeInput(it) }
+        binding.viewEditTextClickable.setOnClickListener { binding.editTextAmount.focusAndShowKeyboard(true) }
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidget.kt:43]: named arg
you can use with(binding)

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            is SwapTokenModel.JupiterToken -> null
+            is SwapTokenModel.UserToken -> availableAmountText(token)
+        }
+
+    private fun availableAmountText(token: SwapTokenModel.UserToken): TextViewCellModel.Raw =
+        TextViewCellModel.Raw(
+            TextContainer(R.string.swap_main_available_amount, tokenAmount(token))
+        )
+
+    private fun tokenAmount(token: SwapTokenModel.UserToken): String =
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:125]: can be as method  of UserToken

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    private val widgetMapper: SwapWidgetMapper,
+    private val buttonMapper: SwapButtonMapper,
+) : BasePresenter<JupiterSwapContract.View>(), JupiterSwapContract.Presenter {
+
+    override fun attach(view: JupiterSwapContract.View) {
+        super.attach(view)
+
+        stateManager.observe()
+            .onEach(::handleFeatureState)
+            .flowOn(dispatchers.io)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:26]: no need to flowOn
and launchIn(this)

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapContract.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package org.p2p.wallet.swap.ui.jupiter.main
+
+import org.p2p.uikit.utils.text.TextViewCellModel
+import org.p2p.wallet.common.mvp.MvpPresenter
+import org.p2p.wallet.common.mvp.MvpView
+import org.p2p.wallet.swap.ui.jupiter.main.widget.SwapWidgetModel
+
+interface JupiterSwapContract {
+    interface View : MvpView {
+
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapContract.kt:10]: empty line

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt [hunk 7]
```diff
@@ -1,63 +1,70 @@
...
 import org.p2p.wallet.utils.viewbinding.viewBinding
 
-class JupiterSwapFragment : BaseFragment(R.layout.fragment_jupiter_swap) {
+class JupiterSwapFragment :
+    BaseMvpFragment<JupiterSwapContract.View, JupiterSwapContract.Presenter>(R.layout.fragment_jupiter_swap),
+    JupiterSwapContract.View {
 
     private val binding: FragmentJupiterSwapBinding by viewBinding()
+    override val presenter: JupiterSwapContract.Presenter
+        get() = TODO("Not yet implemented")
```
Comment (kabramovp2p) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt:27]: TODO() is expected for error or replace with // TODO

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt [hunk 8]
```diff
@@ -1,63 +1,70 @@
...
-            TextViewCellModel.Skeleton(
-                SkeletonCellModel(
-                    height = 16.toPx(),
-                    width = 160.toPx(),
-                    radius = 4f.toPx(),
-                )
-            )
-        )
-
         binding.imageViewSwapTokens.background = shapeDrawable(shapeCircle())
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt:31]: with(binding) { }

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt [hunk 9]
```diff
@@ -1,63 +1,70 @@
...
+    }
+
+    override fun setButtonState(buttonState: SwapButtonState) = with(binding) {
+        when (buttonState) {
+            is SwapButtonState.Disabled -> {
+                buttonError.isInvisible = false
+                sliderSend.isInvisible = true
+                buttonState.text.applyTo(buttonError)
+            }
+            SwapButtonState.Hide -> {
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt:54]: is

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+    private fun handleFeatureState(state: SwapState) {
+        when (state) {
+            SwapState.InitialLoading -> {
+                view?.setFirstTokenWidgetState(widgetMapper.mapWidgetLoading(isTokenA = true))
+                view?.setSecondTokenWidgetState(widgetMapper.mapWidgetLoading(isTokenA = false))
+                view?.setButtonState(SwapButtonState.Hide)
+            }
+            is SwapState.TokenAZero -> {
+                view?.setFirstTokenWidgetState(widgetMapper.mapTokenA(state.tokenA, tokenAmount = null))
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:50]: please use named arg for all variables )

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt (hunk 4 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:125]: UserToken is domain layer model, they not need formate it for ui. don't want overload like org.p2p.core.token.Token

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt (hunk 5 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:26]: why?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt (hunk 7 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapFragment.kt:27]: next pr will fixit it

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt (hunk 5 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:26]: we have our own dispatcher in manager, don't we?

about launchIn - my mistake, sorry

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidgetModel.kt (hunk 2 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/widget/SwapWidgetModel.kt]: create SkeletonExtensions.kt

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt (hunk 5 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt]: yea you right

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            .onEach(::handleFeatureState)
+            .launchIn(this)
+    }
+
+    override fun switchTokens() {
+        stateManager.onNewAction(SwapStateAction.SwitchTokens)
+    }
+
+    override fun onTokenAmountChange(amount: String) {
+        stateManager.onNewAction(SwapStateAction.TokenAAmountChanged(amount.toBigDecimal()))
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:31]: `toBigDecimalOrZero()`

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            state = widgetMapper.mapTokenB(token = state.tokenB, tokenAmount = null)
+        )
+        view?.setButtonState(buttonMapper.mapEnterAmount())
+    }
+
+    private fun handleInitialLoading() {
+        view?.setFirstTokenWidgetState(state = widgetMapper.mapWidgetLoading(isTokenA = true))
+        view?.setSecondTokenWidgetState(state = widgetMapper.mapWidgetLoading(isTokenA = false))
+        view?.setButtonState(buttonState = SwapButtonState.Hide)
+    }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/JupiterSwapPresenter.kt:92]: looks clean, I like it

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+
+    fun mapWidgetLoading(isTokenA: Boolean): SwapWidgetModel {
+        return if (isTokenA) {
+            SwapWidgetModel.Loading(
+                isStatic = false,
+                widgetTitle = TextViewCellModel.Raw(text = TextContainer(R.string.swap_main_you_pay)),
+            )
+        } else {
+            SwapWidgetModel.Loading(
+                isStatic = true,
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:24]: what is static?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        token: SwapTokenModel,
+    ): SwapWidgetModel.Content =
+        SwapWidgetModel.Content(
+            isStatic = true,
+            widgetTitle = swapWidgetToTitle(),
+            currencyName = tokenName(token),
+            amount = textCellSkeleton(
+                height = 20.toPx(),
+                width = 84.toPx(),
+                radius = 6f.toPx(),
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:74]: Are you sure we should have static height for this?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+        return TextViewCellModel.Raw(TextContainer(amountText))
+    }
+
+    private fun tokenName(token: SwapTokenModel): TextViewCellModel.Raw =
+        TextViewCellModel.Raw(
+            TextContainer(
+                when (token) {
+                    is SwapTokenModel.JupiterToken -> token.token.tokenName
+                    is SwapTokenModel.UserToken -> token.token.tokenName
+                }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:104]: that looks weird. 
Just add one fun in `SwapTokenModel` which checks type and returns `tokenName`

File: ui-kit/src/main/kotlin/org/p2p/uikit/utils/text/TextViewCellModel.kt [hunk 16]
```diff
@@ -143,5 +178,12 @@ private data class InitialTextStyle(
...
         hintTextColors = textView.hintTextColors,
+        textViewAutoSizeConfiguration = TextViewAutoSizeConfiguration(
+            autoSizeTextType = textView.autoSizeTextType,
+            autoSizeMinTextSize = textView.autoSizeMinTextSize,
+            autoSizeMaxTextSize = textView.autoSizeMaxTextSize,
+            autoSizeStepGranularity = textView.autoSizeStepGranularity,
+            typedValue = TypedValue.COMPLEX_UNIT_PX,
+        ),
     )
 }
```
Comment (1sakov) [inline ui-kit/src/main/kotlin/org/p2p/uikit/utils/text/TextViewCellModel.kt:189]: I think this class is getting overloaded. 
Maybe, it's better to create a folder for `TextViewCellModel ` and clean here a little bit?

**Review by 1sakov — APPROVED** [review]
Minor comments.
Overall lgtm, feel free to merge after comment fixes

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt (hunk 13 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:24]: Снимок экрана 2023-02-21 в 12 35 32
bottom widget static, disable input, grey background

File: ui-kit/src/main/kotlin/org/p2p/uikit/utils/text/TextViewCellModel.kt (hunk 16 above)
Comment (NikolaiEmelianov) [reply ui-kit/src/main/kotlin/org/p2p/uikit/utils/text/TextViewCellModel.kt:189]: I'll clean it next pr

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt (hunk 14 above)
Comment (NikolaiEmelianov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/main/mapper/SwapWidgetMapper.kt:74]: yes it is less then text size height

**Pull request merged**